### PR TITLE
[match] print certificate info ℹ️

### PR DESCRIPTION
**Pull request by hjanuschka**
this one redoes https://github.com/fastlane/fastlane/pull/5061 - credit goes to @punksta
i had to rework it - as merging/rebasing was to heavy.

bildschirmfoto 2016-11-08 um 15 10 20

### CODE REVIEW

File: match/lib/match/utils.rb [hunk 1]
```diff
@@ -45,6 +45,55 @@ def self.environment_variable_name_profile_name(app_identifier: nil, type: nil)
       (base_environment_variable_name(app_identifier: app_identifier, type: type) + ["profile-name"]).join("_")
     end
 
+    # Fill identity(name) of certificate to env variable
+    def self.fill_environment_cert(params, identity)
```
Comment (milch) [inline match/lib/match/utils.rb]: This seems to be unused, in that case, remove please.

File: match/lib/match/runner.rb [hunk 2]
```diff
@@ -85,6 +85,14 @@ def fetch_certificate(params: nil)
         # Import the private key
         # there seems to be no good way to check if it's already installed - so just install it
         Utils.import(keys.last, params[:keychain_name], password: params[:keychain_password])
+
+        # Get and print info of certificate
+        info = Utils.get_cert_info(cert_path)
+        cert_name = info.select { |key, value| key == 'Common Name' }
+        if cert_name.length > 0
```
Comment (milch) [inline match/lib/match/runner.rb]: Why is this check needed? When would the length be 0?

File: match/lib/match/utils.rb [hunk 3]
```diff
@@ -45,6 +45,55 @@ def self.environment_variable_name_profile_name(app_identifier: nil, type: nil)
...
+      # notAfter={End datetime}
+
+      # removing subject= in first line
+      subject_value = output[0][output[0].index('=') + 2, output[0].length]
+
+      out_array = subject_value.split('/')
+      out_array << output[1]
+      out_array << output[2]
+
+      oppenssl_keys_to_readable_keys = {
```
Comment (milch) [inline match/lib/match/utils.rb]: typo: `oppenssl_keys_to_readable_keys` -> `openssl_keys_to_readable_keys`

File: match/lib/match/utils.rb [hunk 4]
```diff
@@ -45,6 +45,55 @@ def self.environment_variable_name_profile_name(app_identifier: nil, type: nil)
...
+      output = Helper.backticks(command, print: $verbose).split("\n")
+
+      # openssl output:
+
+      # subject= /UID={User ID}/CN={Certificate Name}/OU={Certificate User}/O={Organisation}/C={Country}\n
+      # notBefore={Start datetime}\n
+      # notAfter={End datetime}
+
+      # removing subject= in first line
+      subject_value = output[0][output[0].index('=') + 2, output[0].length]
```
Comment (milch) [inline match/lib/match/utils.rb]: Do `.gsub(/\s*subject=\s*/, "")` and `.gsub("/", "\n")` before running `.split("\n")` instead, that's a bit more clear and doesn't introduce 2 temporaries.

File: match/lib/match/utils.rb [hunk 5]
```diff
@@ -45,6 +45,55 @@ def self.environment_variable_name_profile_name(app_identifier: nil, type: nil)
...
+           'C' => 'Country',
+           'notBefore' => 'Start Datetime',
+           'notAfter' => 'End Datetime'
+       }
+
+      # collect openssl answer to structure [[key1, v1], [key2, v2], ...]
+      key_value_pairs = out_array.map { |x| x.split(/=+/) if x.include? "=" }.compact
+
+      # change keys to readable
+      table_data = key_value_pairs.map { |k, v| [oppenssl_keys_to_readable_keys.fetch(k, k), v] }
```
Comment (milch) [inline match/lib/match/utils.rb]: We don't need to introduce temporaries here, with the changes from above you can do:

```ruby
       output.map { |x| x.split(/=+/) if x.include? "=" }
                  .compact
                  .map { |k, v| [oppenssl_keys_to_readable_keys.fetch(k, k), v] }
```

**Review by milch — CHANGES_REQUESTED** [review]
Just tried it, I really like this change, thanks! 🚀  Needs some simplification, then we can merge.

**Comment by hjanuschka** [comment]
@milch feedback addressed.

about the `cert_info.length > 0` check on `common name` - i took it from the original PR - i think its because of "in X.509 the Common Name is not mandatory." - but i think its ok to skip/remove that check, as in ios world - common name is always set.

  * rebased current master

**Review by milch — APPROVED** [review]

**Comment by milch** [comment]
@hjanuschka Alright, this looks good to me then. Thanks for your contribution! 🚀 

PS: In the future, when addressing reviews and/or rebasing on master, please don't `--amend` and force push. It makes it easier to just quickly scan over just the requested changes, instead of having to go through all the code again. We squash anyway before merging 😉

**Comment by hjanuschka** [comment]
@milch - ok thx for the info, will do so in the future - just a bad habbit (from back in the days when it was not possible to squash) to submit only one commit!

**Comment by KrauseFx** [comment]
Is the country always US?

**Comment by hjanuschka** [comment]
@KrauseFx seems to be always US - not sure why - either apple or sigh (in the crt) ?

**Comment by hjanuschka** [comment]
@KrauseFx  infact i think its defaulting to US - because  fastlane/spaceship - does not set a country code.


https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/portal/certificate.rb#L195 the country should be included in the subject of the CSR.


sample on how to set country code:
https://gist.github.com/mitfik/1922961#file-ruby_csr_example-rb-L31

**Pull request merged**